designs/0031-deprecation-schedule.md

```diff
@@ -0,0 +1,254 @@
+   removing deprecations, emitting warnings to the user is simply annoying and
+   nonproductive.
+
+   We instead decide to strongly commit to backwards compatibility in all
+   cases - no new language keywords, the preservation of current syntax ad
+   infinitum, etc. This has benefits to the end user - a Stan program written
+   today would still work (assuming we were actually able to meet this high
+   standard) on a version of Stan released years from now. The downside is that
+   the codebase would bloat over time, and we will run into more and more
+   instances where the correct behavior or error message is indiscernible in
+   practice. Tools which build around Stan in particular have trouble with our
+   treatment of some things as valid identifiers and valid keywords at the same
+   time - syntax highlighting especially suffers:
+   ```stan
+       real<lower=0> x; // notice lower is highlighted
+       real<offset=0> y; // notice offset is not
+       int offset; // this is (partially) the reason why
+   ```
+
+# Prior art
+[prior-art]: #prior-art
+
+- The original inspiration for this schedule comes from the Python library
+  NumPy, with their deprecation policy laid out in
+  [NEP-23](https://numpy.org/neps/nep-0023-backwards-compatibility.html#implementing-deprecations-and-removals).
+  The NumPy project provides a similar example of a large open-source code base
+  whose minor version number has now exceeded 20 (meaning a large number of
+  changes since the last 'breaking'/major version).
+- Python itself defines a similar policy in
+  [PEP-387](https://www.python.org/dev/peps/pep-0387/) which requires warnings
+  emitted for 2 minor versions or 1 major version (e.g, a feature deprecated in
+  3.10 could be removed in 3.12 **or** 4.0)
+- The Java language clarified its [deprecation
+  policy](https://docs.oracle.com/javase/9/core/enhanced-deprecation1.htm#JSCOR-GUID-23B13A9E-2727-42DC-B03A-E374B3C4CE96)
+  in Java 9 (which is technically a minor version, 1.9, for the JDK). The main
+  motivation for clarification was the response of users being unsure if a
+  deprecated feature would ever actually be removed. There was a
+  [study](https://doi.org/10.1007/S10664-019-09713-W) of this change published
+  in 2019. It contains the following relevant passage and some interesting
+  discussion of user reactions to a deprecation warning.
+  > By not removing deprecated features from an API after a transition period
+  > has passed, API producers and the Java JDK developers themselves have
+  > cheapened the meaning of deprecation.
+- https://semver.org/ - the description of proper 'semantic versioning'. In the
+  strictest interpretation, Stan already deviates from this scheme with some
+  regularity, but it is a common understanding of the major.minor.patch
+  versioning.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Above I set the term of deprecations at 3 minor versions/1 year. This may be
```

Comment:
Let's just fix to 3 minor versions or 1 major version like Python.  We don't want every deprecation to involve a debate.

Reply:
Even Python includes the line

> It's fine to wait more than two releases.

when describing their policy. I agree that a priori it seems strange to think of when something would be fine to remove, but wanting more than 3 versions.

Reply:
Another way of looking at this is the whole spec is aimed at preventing the "let's just wait" approach to deprecation.

Let me ask this another way.  When would we want to wait more than 3 releases?

Reply:
That's fair. I've updated it other places, but it is in the "Unresolved questions" portion in part because I wonder if 3 is the best number for a hard and fast rule or if we would prefer 2 minor versions (so you have the deprecating version, a buffer version, and the removal version).